hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractDistributedObject.java

```diff
@@ -64,7 +66,8 @@ public final void destroy() {
         if (preDestroy()) {
             NodeEngine engine = getNodeEngine();
             ProxyService proxyService = engine.getProxyService();
-            proxyService.destroyDistributedObject(getServiceName(), getDistributedObjectName());
+            UUID source = engine.getLocalMember().getUuid();
```

Comment:
what if i call this method from a client message task? In that case event can contain wrong source uuid? Can we prevent this happening? If not this seems a weakness point.

Reply:
In client message tasks, we normally try NOT to use server side proxy calls but make the related operation calls directly. Otherwise it is a bad practice. But having said that, the method itself does not prevent it. 

I can put this assert check which uses reflection at the beginning of destroy method, would that be OK? :

```
    public final void destroy() {
        assert(!isCalledFromClientMessageTask());

        if (preDestroy()) {
            NodeEngine engine = getNodeEngine();
            ProxyService proxyService = engine.getProxyService();
            UUID source = engine.getLocalMember().getUuid();
            proxyService.destroyDistributedObject(getServiceName(), getDistributedObjectName(), source);
            postDestroy();
        }
    }

    private boolean isCalledFromClientMessageTask() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        for (StackTraceElement e : stackTraceElements) {
            if (e.getClassName().contains("MessageTask")) {
                return true;
            }
        }
        return false;
    }

```

Reply:
Seemed a costly solution. Since currently we don't have such problem, maybe we can only add a short comment which mentions the potential issue.

Reply:
Added a note at the beginning of the method.